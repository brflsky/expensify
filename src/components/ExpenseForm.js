import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();
class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense.description ? props.expense.description : '',
      notes: props.expense.notes ? props.expense.notes : '',
      amount: props.expense.amount ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense.createdAt ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  onNotesChange = (e) => {
    const notes = e.target.value;
    this.setState(() => ({ notes }));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  onRemoveClick = (e) => {
    e.preventDefault();
    this.props.onRemove();
  }

  onDateChange = (createdAt) => {
    if (createdAt)
      this.setState(() => ({ createdAt }));
  }
  onCalenderFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  }
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please enter required values' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        notes: this.state.notes,
        createdAt: this.state.createdAt.valueOf()
      });
    }
  }
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          id="description"
          type="text"
          className="text-input"
          placeholder="Expense Title"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          placeholder="Amount"
          className="text-input"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calenderFocused}
          onFocusChange={this.onCalenderFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className="textarea"
          placeholder="Add some notes here"
          onChange={this.onNotesChange}
          value={this.state.notes}
        />
        <div className="header__content">
          <button className="button">Save Expense</button>
          {this.props.expense.id && <button className="button button--second" onClick={this.onRemoveClick}>Remove Expense</button>}
        </div>
      </form>
    );
  }
}

export default ExpenseForm;
