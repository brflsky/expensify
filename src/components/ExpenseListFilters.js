import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setSortAsc, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null,
    sortAsc: true
  };


  onDateChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  onFocusChange = (calenderFocused) => {
    this.setState(() => ({ calenderFocused }));
  }

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else {
      this.props.sortByAmount();
    }
  }

  onToggleSort = (e) => {
    const sortAsc = !this.state.sortAsc;
    this.setState((state) => ({
      ...state,
      sortAsc
    }));
    if (sortAsc) {
      this.props.setSortAsc(true);
    } else {
      this.props.setSortAsc(false);
    }
  }

  onSortAscDescChange = (e) => {
    if (e.target.value === 'true') {
      this.props.setSortAsc(true);
    } else {
      this.props.setSortAsc(false);
    }
  }

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input className="text-input" placeholder="Search Expenses" type="text" value={this.props.filters.text} onChange={this.onTextChange} />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              onChange={this.onSortChange}
              value={this.props.filters.sortBy}
              placeholder="Add Expenses"
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
          <p id="sortAsc" onClick={this.onToggleSort} className="sortAsc">{this.state.sortAsc ? '\u21E7' : '\u21E9'}</p>
           {/** {this.state.sortAsc ? <p  className="">(&#11014)</p> : <p  className="">(&#11015)</p>}
             <select
              className="select"
              onChange={this.onSortAscDescChange}
              value={this.props.filters.sortAsc ? 'true' : 'false'}
            >
              <option value="true">Asc</option>
              <option value="false">Desc</option>
            </select> */}
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDateChange}
              focusedInput={this.state.calenderFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>

      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => {
  return {
    setStartDate: (date) => dispatch(setStartDate(date)),
    setEndDate: (date) => dispatch(setEndDate(date)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setSortAsc: (e) => dispatch(setSortAsc(e)),
    setTextFilter: (e) => dispatch(setTextFilter(e))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
