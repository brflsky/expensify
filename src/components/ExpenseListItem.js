import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import preventClickthrough from 'react-prevent-clickthrough';
import { startRemoveExpense } from '../actions/expenses';

// export const ExpenseListItem = (props) => {
//   // console.log('props', props);
//   return (
//     <Link className="list-item" to={`/edit/${props.id}`} >
//       <div>
//         <h3 className="list-item__title">{props.index}. &nbsp; {props.description}</h3>
//         <span className="list-item__sub-title">{moment(props.createdAt).format('Do-MMMM-YYYY')}</span>
//       </div>
//       <div>
//       <h3 className="list-item__data">{numeral(props.amount / 100).format('$0,0.00')}</h3>
//       <button className="list-item__button" onClick={(e) => { props.dispatch(startRemoveExpense(props, e)); }}>remove</button>
//       </div>
//     </Link>
//   );
// };

export class ExpenseListItem extends React.Component {
  onRemove = (e) => {
    e.stopPropagation();
    //e.nativeEvent.stopImmediatePropagation();
    //e.stopPropagation();
    this.props.dispatch(startRemoveExpense(this.props));
  }
  onChoose = (e) => {
    this.props.history.push(`/edit/${this.props.id}`);
  }
  render() {
    return (
      <div className="list-item" onClick={this.onChoose} >
        <div>
          <h3 className="list-item__title">{this.props.index}. &nbsp; {this.props.description}</h3>
          <span className="list-item__sub-title">{moment(this.props.createdAt).format('Do-MMMM-YYYY')}</span>
        </div>
        <div>
          <h3 className="list-item__data">{numeral(this.props.amount / 100).format('$0,0.00')}</h3>
          <button className="list-item__button" onClick={this.onRemove}>remove</button>
        </div>
      </div>
    );
  }
}

export default connect()(ExpenseListItem);
