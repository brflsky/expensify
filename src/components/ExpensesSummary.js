import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import select from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => (
  <div className="page-header">
    <div className="content-container">
      <h1 className="page-header__title">
        Viewing <span>{props.expenses.length}</span>. Total cost: <span>{numeral(getExpensesTotal(props.expenses) / 100).format('$0,0.00')}</span>
      </h1>
      <div className="page-header__actions">
        <Link className="button" to="/create" >Add Expense</Link>
      </div>
    </div>
  </div>
);


const mapStateToProps = (state) => ({
  expenses: select(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);
