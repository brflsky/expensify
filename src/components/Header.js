import React from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { startLogout } from '../actions/auth';

export const Header = ({ logout }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/dashboard" ><h1>Expensify</h1></Link>
        <button className="button button--link" onClick={logout}>Logout</button>
      </div>
    </div>
  </header>

);

const mapDispatchToProps = (dispatch) => ({
  logout: () => dispatch(startLogout())
});

export default withRouter(connect(null, mapDispatchToProps)(Header));
