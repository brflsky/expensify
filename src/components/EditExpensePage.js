import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';


export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    //console.log('updated',expense);
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }
  onClick = () => {
    this.props.removeExpense (this.props.expense);
    this.props.history.push('/');
  }
  render() {
    return  (
      <div>
        <ExpenseForm 
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onClick}
        >X</button>
      </div>
    );
  }
}


const mapStateToProps = (state, props) =>{
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
};
const mapsDispatchToProps = (dispatch) => {
  return {
    editExpense: (id, expense) => dispatch(editExpense(id,expense)),
    removeExpense: expense => dispatch(removeExpense(expense))
  }
}

export default connect(mapStateToProps, mapsDispatchToProps)(EditExpensePage);  