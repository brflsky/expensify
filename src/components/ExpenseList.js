import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import selectedExpenseList from '../selectors/expenses';

export const ExpenseList = props => (
  <div>
      {props.expenses.length > 0 ?  props.expenses.map((expense, index) => (
      <ExpenseListItem 
        key={expense.id}
        id={expense.id}
        index={index+1}
        description={expense.description}
        amount={expense.amount}
        createdAt={expense.createdAt}
      />
    )) : 
    <p>Nothing to show... ad something</p>
  }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectedExpenseList(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);

// const ConnectedExpenseList = connect((state) => {
//   return {
//     expenses: state.expenses
//   }
// })(ExpenseList);

// export default ConnectedExpenseList;