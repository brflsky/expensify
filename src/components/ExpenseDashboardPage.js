import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpansifyDashboardPage = (props) => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList history={props.history} />
  </div>
);

export default ExpansifyDashboardPage;
