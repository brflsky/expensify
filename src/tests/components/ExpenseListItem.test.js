import React from 'react';
import { shallow } from 'enzyme';
import {ExpenseListItem} from '../../components/ExpenseListItem';
import {expenses} from '../fixtures';

test('Shold render single expense ', () => {
const wrapper = shallow(<ExpenseListItem index={0} { ...expenses[0] }/>);
expect(wrapper).toMatchSnapshot();

});