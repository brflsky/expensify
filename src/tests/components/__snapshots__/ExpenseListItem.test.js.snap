// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shold render single expense  1`] = `
<div>
  <p>
    0
    .  
    <Link
      replace={false}
      to="/edit/0"
    >
      Gum
    </Link>
     
    $1.99
     
    1st-January-1970
      
    <button
      onClick={[Function]}
    >
      X
    </button>
  </p>
</div>
`;
